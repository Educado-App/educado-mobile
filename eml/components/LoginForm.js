import React from "react";
import {StyleSheet, Pressable, Text, Image, View, Dimensions, TextInput} from "react-native";

const {width, height} = Dimensions.get('window');

import {useNavigation} from "@react-navigation/native";
import {useRoute} from "@react-navigation/native";

export default function LoginForm(props) {
    const navigation = useNavigation();
    const route = useRoute().name;

    return (
        <View style ={styles.container}>
            <View style ={StyleSheet.absoluteFill}>
                <View style ={styles.textLogoContainer}>
                    <Text>Educado...</Text>
                </View>

               {/* <View style={styles.closeButtonContainer}>


                <View style={styles.closeButtonContainer}>

                    <Text>X</Text>

                </View>*/}
            </View>
            <View style={styles.bottomContainer}>
                <View style={styles.formInputContainer}>
                    <TextInput placeholder="Phone Number" placeholderTextColor="black" style={styles.textInput} />
                    <TextInput placeholder="Password" placeholderTextColor="black"  style={styles.textInput} />
                    <View style={styles.formButton}>
                        <Text style={styles.buttonText}> LOG IN</Text>
                    </View>
                </View>
            </View>
        </View>
    );
}


const styles = StyleSheet.create({
        container: {
            flex: 1,
            justifyContent: 'flex-center',
            backgroundColor: 'rgba(0, 173, 35,0.8)'
        },
        button: {
            backgroundColor : 'rgba(123,104,238,0.8)',
            height : 55,
            alignItems : 'center',
            justifyContent : 'center',
            borderRadius: 35,
            marginHorizontal : 20,
            marginVertical: 10,
            borderWidth: 1,
            borderColor: 'white'
        },
        buttonText:{
            fontSize: 20,
            fontWeight: '600',
            color: 'black',
            letterSpacing: 0.5

        },
        bottomContainer:{
            justifyContent: 'center',
            height: height,
        },
        textInput: {
            height: 50,
            borderWidth: 1,
            borderColor: 'rgba(0,0,0, 0.2)',
            marginHorizontal: 20,
            marginVertical: 10,
            borderRadius: 25,
            paddingLeft: 10
        },
        formButton: {
            backgroundColor : 'white',
            height : 55,
            alignItems : 'center',
            justifyContent : 'center',
            borderRadius: 35,
            marginHorizontal : 20,
            marginVertical: 10,
            borderWidth: 1,
            borderColor: 'white',
            shadowColor: "#000",
            shadowOffset: {
                width: 0,
                height: 4,
            },
            shadowOpacity: 0.25,
            shadowRadius: 3.84,
            elevation: 5,

        },
        formInputContainer:{
            marginBottom: 70
        },
       /* closeButtonContainer:{
            height: 40,
            width: 40,
            justifyContent: 'center',
            alignSelf: 'center',
            shadowColor: "#000",
            shadowOffset: {
                width: 0,
                height: 5,
            },
            shadowOpacity: 0.34,
            shadowRadius: 6.27,
            elevation: 1,
            backgroundColor: 'white',
            alignItems: 'center',
            borderRadius: 20
        },*/
        textLogoContainer: {
            marginHorizontal : 20,
            marginVertical: 75,
            fontSize: 30,
            fontWeight: '600',
            color: 'green',
            letterSpacing: 0.5,
            height: 50,
            width: 1000,
            justifyContent: 'center'
        }
});
