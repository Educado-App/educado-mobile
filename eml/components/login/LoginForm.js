import React, { useState } from 'react';
import { View } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { loginUser } from '../../api/userApi';
import FormTextField from '../general/forms/FormTextField';
import FormButton from '../general/forms/FormButton';
import PasswordEye from '../general/forms/PasswordEye';
import ResetPassword from './ResetPassword';
import FormFieldAlert from '../general/forms/FormFieldAlert';
import { removeEmojis } from '../general/Validation';
import Text from '../general/Text';
import ShowAlert from '../general/ShowAlert';


// Services
import { setUserInfo, setJWT } from '../../services/StorageService';
import AsyncStorage from '@react-native-async-storage/async-storage';

/**
 * Login form component for login screen containing email and password input fields and a login button.
 * @returns {React.Element} Component for logging in (login screen)
 */
export default function LoginForm() {

	const navigation = useNavigation();
	const [email, setEmail] = useState('');
	const [password, setPassword] = useState('');
	const [modalVisible, setModalVisible] = useState(false);
	const [passwordAlert, setPasswordAlert] = useState('');
	const [emailAlert, setEmailAlert] = useState('');
	// State variable to track password visibility
	const [showPassword, setShowPassword] = useState(false);

	/**
   * Logs user in with the entered credentials 
   * @param {String} email Email user tries to login with
   * @param {String} password Password user tries to login with
   */
	async function login(email, password) {

		//Reset alerts
		setEmailAlert('');
		setPasswordAlert('');

		//The Object must be hashed before it is sent to backend (before loginUser() is called)
		//The Input must be conditioned (at least one capital letter, minimum 8 letters and a number etc.)
		const obj = {
			email: email,
			password: password,
		};


		// Await the response from the backend API for login
		await loginUser(obj).then(async (response) => {
			// Set login token in AsyncStorage and navigate to home screen
			await setJWT(response.accessToken);
			await setUserInfo(response.userInfo);
			await AsyncStorage.setItem('loggedIn', 'true');
			navigation.navigate('HomeStack');
		}).catch((error) => {
			switch (error?.error?.code) {
			case 'E0004':
				// No user exists with this email!
				setEmailAlert('Não existe nenhum usuário com este email!');
				break;

			case 'E0105':
				// Password is incorrect!
				setPasswordAlert('Senha incorreta!');
				break;

			case 'E0003':
				// Error connecting to server!
				ShowAlert('Erro de conexão com o servidor!');
				break;

				// TODO: What error should we give here instead? Unknown error? 
			default: // Errors not currently handled with specific alerts
				ShowAlert('Erro desconhecido!');
			}
		});
	}


	// Function to close the reset password modal
	const closeModal = () => {
		setModalVisible(false);
	};


	// Function to toggle the password visibility state
	const toggleShowPassword = () => {
		setShowPassword(!showPassword);
	};

	return (
		<View>
			<View className="mb-6">
				<FormTextField
					testId="emailInput"
					placeholder="Insira sua e-mail"
					onChangeText={(email) => setEmail(email)}
					label="E-mail"
					required={true}
					keyboardType="email-address"
				/>
				<FormFieldAlert testId="emailAlert" label={emailAlert} />
			</View>


			<View className="relative mb-6">
				<FormTextField
					testId="passwordInput"
					placeholder="Insira sua senha" // Type your password
					value={password}
					onChangeText={(inputPassword) => {
						setPassword(removeEmojis(inputPassword, password));
					}}
					label="Senha" // Password
					required={true}
					secureTextEntry={!showPassword}
				/>
				<PasswordEye
					testId="passwordEye"
					showPasswordIcon={showPassword}
					toggleShowPassword={toggleShowPassword}
				/>
				<FormFieldAlert testId="passwordAlert" label={passwordAlert} />
			</View>

			<View>
				{/* TODO: tilføj onPress til nedenstående; reset password */}
				<Text
					className={'text-right underline text-base text-projectBlack mb-15'}
					onPress={() => setModalVisible(true)}
				>
					{/* reset your password? */}
          Esqueceu a senha?
				</Text>
			</View>
			{/* Enter */}
			<FormButton
				testId="loginButton"
				onPress={() => login(email, password)}
				disabled={!(password.length > 0 && email.length > 0)}
				className='mt-4'
			>
        Entrar
			</FormButton>
			<View className="pt-10">
				<ResetPassword
					className={(!modalVisible ? 'hidden' : '')}
					modalVisible={modalVisible}
					onModalClose={closeModal}
					testId="resetPasswordModal"
					// Reset password
					title="Redefinir senha"
				/>
			</View>
		</View>
	);
}
