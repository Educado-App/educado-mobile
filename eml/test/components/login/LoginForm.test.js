import renderer from 'react-test-renderer';
import React from 'react';
import LoginForm from '../../../components/login/LoginForm';
import AsyncStorage from '@react-native-async-storage/async-storage';

let loginForm;
const LOGIN_TOKEN = '@loginToken';
const USER_INFO = '@userInfo';
const testJwt = 'token';

const testUserInfo = {
  id: '123',
  firstName: 'Is',
  lastName: 'User',
  email: 'is@user.com',
};

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigate: jest.fn(),
  }),
}));

jest.useRealTimers('legacy');

jest.mock('../../../api/userApi', () => ({
  loginUser: jest.fn(async ({ email, password }) => {

    if (email === testUserInfo.email && password === 'password123') {
      return Promise.resolve({ accessToken: testJwt, userInfo: testUserInfo });
    } else if (email !== testUserInfo.email) {
      return Promise.reject({ error: { code: 'E0004' } });
    } else if (password !== 'password123') {
      return Promise.reject({ error: { code: 'E0105' } });
    }
  })
}));

beforeEach(async () => {
  await renderer.act(async () => {
    loginForm = renderer.create(<LoginForm />);
  });
});

afterEach(() => {
  loginForm = null;
});

describe('login', () => {

  test('Check LoginForm renders correctly', async () => {
    const tree = loginForm.toJSON();
    expect(tree).toMatchSnapshot();
  });


  test('Login function when the login button is pressed', async () => {

    // Import the loginUser function from the mocked userApi module
    const loginUser = require('../../../api/userApi').loginUser;

    // Mock the loginUser function to resolve with a dummy response
    //loginUser.mockResolvedValue({ accessToken: "dummyToken" });


    const emailInput = loginForm.root.findByProps({ testId: 'emailInput' });
    const passwordInput = loginForm.root.findByProps({ testId: 'passwordInput' });
    const loginButton = loginForm.root.findByProps({ testId: 'loginButton' });

    // Fill in email and password and press the login button
    await renderer.act(async () => {
      emailInput.props.onChangeText('test@example.com');
      passwordInput.props.onChangeText('password123');
    });

    await renderer.act(async () => {
      loginButton.props.onPress();
    });

    // Ensure login api function was called with the correct arguments
    expect(loginUser).toHaveBeenCalledWith({
      email: 'test@example.com',
      password: 'password123',
    });
  });
});

describe('Form behaviour', () => {

  test('Check that password visibility is toggled correctly', async () => {
    const passwordEye = loginForm.root.findByProps({ testId: 'passwordEye' });
    const passwordInput = loginForm.root.findByProps({ testId: 'passwordInput' });
    expect(passwordEye.props.showPasswordIcon).toBe(false);
    expect(passwordInput.props.secureTextEntry).toBe(true);
    await renderer.act(async () => {
      passwordEye.props.toggleShowPassword();
    });
    expect(passwordEye.props.showPasswordIcon).toBe(true);
    expect(passwordInput.props.secureTextEntry).toBe(false);
    await renderer.act(async () => {
      passwordEye.props.toggleShowPassword();
    });
    expect(passwordEye.props.showPasswordIcon).toBe(false);
    expect(passwordInput.props.secureTextEntry).toBe(true);
  });


  test('Check that login button is disabled when email or password is empty', async () => {
    const emailInput = loginForm.root.findByProps({ testId: 'emailInput' });
    const passwordInput = loginForm.root.findByProps({ testId: 'passwordInput' });
    const loginButton = loginForm.root.findByProps({ testId: 'loginButton' });

    // Button disabled when fields are empty
    expect(loginButton.props.disabled).toBe(true);

    // Button not disabled when fields are filled
    await renderer.act(async () => {
      emailInput.props.onChangeText('thej.dk');
      passwordInput.props.onChangeText('testing123');
    });
    expect(loginButton.props.disabled).toBe(false);

    // Button disabled when email is empty
    await renderer.act(async () => {
      emailInput.props.onChangeText('');
      passwordInput.props.onChangeText('testing123');
    });
    expect(loginButton.props.disabled).toBe(true);

    // Button not disabled when fields are filled
    await renderer.act(async () => {
      emailInput.props.onChangeText('thej.dk');
      passwordInput.props.onChangeText('testing123');
    });
    expect(loginButton.props.disabled).toBe(false);

    // Button disabled when password is empty
    await renderer.act(async () => {
      emailInput.props.onChangeText('thej.dk');
      passwordInput.props.onChangeText('');
    });
    expect(loginButton.props.disabled).toBe(true);
  });


  test('Password field filters out emojis', async () => {
    const passwordInput = loginForm.root.findByProps({ testId: 'passwordInput' });
    await renderer.act(async () => {
      passwordInput.props.onChangeText('testing123');
    });
    expect().toBeFalsy();
    await renderer.act(async () => {
      passwordInput.props.onChangeText('testing123ðŸ¤”');
    });
    expect(passwordInput.props.value).toBe('testing123');
  });


  test('Check that modal opens when clicking on \'forgot password\'', async () => {
    const resetPasswordModal = loginForm.root.findByProps({ testId: 'resetPasswordModal' });
    expect(resetPasswordModal.props.className).toBe('hidden');
  });

  test('Test email alert', async () => {
    const emailInput = loginForm.root.findByProps({ testId: 'emailInput' });
    const passwordInput = loginForm.root.findByProps({ testId: 'passwordInput' });
    const emailAlert = loginForm.root.findByProps({ testId: 'emailAlert' });
    const loginButton = loginForm.root.findByProps({ testId: 'loginButton' });

    await renderer.act(async () => {
      await emailInput.props.onChangeText('not@user.com');
      await passwordInput.props.onChangeText('testpassword123');
      loginButton.props.onPress();
    }).then(() => {
      expect(emailAlert.props.label).not.toBe('');
      //expect(emailAlert.props.label).not.toBe((""));
    });
  });

  test('Test password alert', async () => {
    const emailInput = loginForm.root.findByProps({ testId: 'emailInput' });
    const passwordInput = loginForm.root.findByProps({ testId: 'passwordInput' });
    const passwordAlert = loginForm.root.findByProps({ testId: 'passwordAlert' });
    const loginButton = loginForm.root.findByProps({ testId: 'loginButton' });

    await renderer.act(async () => {
      await emailInput.props.onChangeText('is@user.com');
      await passwordInput.props.onChangeText('wrongpassword');
      loginButton.props.onPress();
    }).then(() => {
      expect(passwordAlert.props.label)
        .not.toBe((''));
    });
  });
});

describe('localstorage', () => {

  it('should save jwt in localstorage', async () => {
    await renderer.act(async () => {
      await emailInput.props.onChangeText(testUserInfo.email);
      await passwordInput.props.onChangeText('password123');
      loginButton.props.onPress();
    }).then(() => {
      expect(AsyncStorage.setItem).toHaveBeenCalledWith(LOGIN_TOKEN, testJwt);
    });
  });


  it('should save userInfo in localstorage', async () => {
    await renderer.act(async () => {
      await emailInput.props.onChangeText(testUserInfo.email);
      await passwordInput.props.onChangeText('password123');
      loginButton.props.onPress();
    }).then(() => {
      expect(AsyncStorage.setItem).toHaveBeenCalledWith(USER_INFO, testUserInfo);
    });
  });
});