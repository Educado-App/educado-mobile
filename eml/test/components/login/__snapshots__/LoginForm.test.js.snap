// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login Check LoginForm renders correctly 1`] = `
<View>
  <View
    style={
      [
        {
          "marginBottom": 24,
        },
      ]
    }
  >
    <View>
      <View
        style={
          [
            {
              "display": "flex",
            },
            {
              "flexDirection": "row",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "Montserrat-Regular",
              },
              {
                "color": "#383838",
              },
              {
                "fontSize": 16,
              },
              [
                {
                  "marginLeft": 8,
                },
                {
                  "fontSize": 12,
                  "lineHeight": 16,
                },
              ],
            ]
          }
        >
          E-mail
        </Text>
        <Text
          style={
            [
              {
                "fontFamily": "Montserrat-Regular",
              },
              {
                "color": "#383838",
              },
              {
                "fontSize": 16,
              },
              [
                {
                  "marginLeft": 4,
                },
                {
                  "fontSize": 12,
                  "lineHeight": 16,
                },
                {
                  "color": "#FF4949",
                },
              ],
            ]
          }
        >
          *
        </Text>
      </View>
      <View>
        <TextInput
          autoComplete="off"
          keyboardType="email-address"
          onChangeText={[Function]}
          passwordGuidelines={false}
          placeholder="Insira sua e-mail"
          secureTextEntry={false}
          style={
            [
              {
                "paddingBottom": 4,
                "paddingTop": 4,
              },
              {
                "paddingLeft": 10,
              },
              {
                "backgroundColor": "#FFFFFF",
              },
              {
                "borderBottomLeftRadius": 8,
                "borderBottomRightRadius": 8,
                "borderTopLeftRadius": 8,
                "borderTopRightRadius": 8,
              },
            ]
          }
        />
      </View>
    </View>
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          {
            "alignItems": "center",
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "Montserrat-Regular",
            },
            {
              "color": "#383838",
            },
            {
              "fontSize": 16,
            },
            [
              {
                "fontSize": 12,
                "lineHeight": 16,
              },
              {
                "color": "#FF4949",
              },
              {
                "marginLeft": 8,
                "marginRight": 8,
              },
            ],
          ]
        }
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "position": "relative",
        },
        {
          "marginBottom": 24,
        },
      ]
    }
  >
    <View>
      <View
        style={
          [
            {
              "display": "flex",
            },
            {
              "flexDirection": "row",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "Montserrat-Regular",
              },
              {
                "color": "#383838",
              },
              {
                "fontSize": 16,
              },
              [
                {
                  "marginLeft": 8,
                },
                {
                  "fontSize": 12,
                  "lineHeight": 16,
                },
              ],
            ]
          }
        >
          Senha
        </Text>
        <Text
          style={
            [
              {
                "fontFamily": "Montserrat-Regular",
              },
              {
                "color": "#383838",
              },
              {
                "fontSize": 16,
              },
              [
                {
                  "marginLeft": 4,
                },
                {
                  "fontSize": 12,
                  "lineHeight": 16,
                },
                {
                  "color": "#FF4949",
                },
              ],
            ]
          }
        >
          *
        </Text>
      </View>
      <View>
        <TextInput
          autoComplete="off"
          keyboardType="default"
          onChangeText={[Function]}
          passwordGuidelines={false}
          placeholder="Insira sua senha"
          secureTextEntry={true}
          style={
            [
              {
                "paddingBottom": 4,
                "paddingTop": 4,
              },
              {
                "paddingLeft": 10,
              },
              {
                "backgroundColor": "#FFFFFF",
              },
              {
                "borderBottomLeftRadius": 8,
                "borderBottomRightRadius": 8,
                "borderTopLeftRadius": 8,
                "borderTopRightRadius": 8,
              },
            ]
          }
          value=""
        />
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "position": "absolute",
          },
          {
            "right": 0,
          },
          {
            "top": 8,
          },
          {
            "paddingBottom": 12,
            "paddingLeft": 12,
            "paddingRight": 12,
            "paddingTop": 12,
          },
        ]
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "gray",
              "fontSize": 24,
            },
            undefined,
            {
              "fontFamily": "material-community",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        ó°ˆˆ
      </Text>
    </View>
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          {
            "alignItems": "center",
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "Montserrat-Regular",
            },
            {
              "color": "#383838",
            },
            {
              "fontSize": 16,
            },
            [
              {
                "fontSize": 12,
                "lineHeight": 16,
              },
              {
                "color": "#FF4949",
              },
              {
                "marginLeft": 8,
                "marginRight": 8,
              },
            ],
          ]
        }
      />
    </View>
  </View>
  <View>
    <Text
      onPress={[Function]}
      style={
        [
          {
            "fontFamily": "Montserrat-Regular",
          },
          {
            "color": "#383838",
          },
          {
            "fontSize": 16,
          },
          [
            {
              "textAlign": "right",
            },
            {
              "textDecorationLine": "underline",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            {
              "color": "#383838",
            },
          ],
        ]
      }
    >
      Esqueceu a senha?
    </Text>
  </View>
  <View>
    <View
      accessibilityState={
        {
          "disabled": true,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#5ECCE9",
          "borderBottomLeftRadius": 8,
          "borderBottomRightRadius": 8,
          "borderTopLeftRadius": 8,
          "borderTopRightRadius": 8,
          "marginTop": 16,
          "opacity": 0.5,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
        }
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "Montserrat-Regular",
            },
            {
              "color": "#383838",
            },
            {
              "fontSize": 16,
            },
            [
              {
                "textAlign": "center",
              },
              {
                "fontFamily": "Montserrat-Bold",
              },
              {
                "fontSize": 16,
              },
              {
                "color": "#FFFFFF",
              },
            ],
          ]
        }
      >
        Entrar
      </Text>
    </View>
  </View>
  <View
    style={
      [
        {
          "paddingTop": 40,
        },
      ]
    }
  >
    <Modal
      animationType="slide"
      hardwareAccelerated={false}
      style={
        [
          {
            "borderBottomWidth": 8,
            "borderLeftWidth": 8,
            "borderRightWidth": 8,
            "borderTopWidth": 8,
          },
        ]
      }
      visible={false}
    />
  </View>
</View>
`;
